File: frontend-web/src/services/gameService.ts
import { api } from './api';
import { Game } from '../types/game.types';
import { PaginatedResponse, PaginationParams } from '../types/api.types';

export interface Award {
  id: number;
  name: string;
  slug: string;
  year: number;
  category: string;
  relevance?: number;
  website?: string;
}

export interface GameWithAwards extends Game {
  awards?: Award[];
  developer?: { id: number; name: string; slug: string };
  platforms?: Array<{ id: number; name: string; slug: string }>;
}

export interface UpcomingGameWithCountdown extends Game {
  developer?: { id: number; name: string; slug: string };
  platforms?: Array<{ id: number; name: string; slug: string }>;
  countdown?: {
    release_date: string;
    days_until_release: number;
    is_released: boolean;
    countdown_label: string;
  };
}

export const gameService = {
  async getAllGames(params?: PaginationParams & { 
    search?: string;
    release_status?: string;
    availability_status?: string;
    year?: number;
    sort?: string;
    order?: string;
    awards?: boolean;
  }): Promise<PaginatedResponse<Game>> {
    const response = await api.get<PaginatedResponse<Game>>('/games', { params });
    return response.data;
  },

  async getGameById(id: number): Promise<Game> {
    const response = await api.get(`/games/${id}`);
    return response.data.data.game;
  },

  async searchGames(query: string, params?: PaginationParams): Promise<PaginatedResponse<Game>> {
    const response = await api.get<PaginatedResponse<Game>>('/games/search', {
      params: { q: query, ...params },
    });
    return response.data;
  },

  async getGotyGames(params?: PaginationParams): Promise<any> {
    const response = await api.get('/games/goty', { params });
    return response.data;
  },

  async getUpcomingReleases(params?: PaginationParams): Promise<PaginatedResponse<Game>> {
    const response = await api.get<PaginatedResponse<Game>>('/games/upcoming-releases', { params });
    return response.data;
  },

  async getUpcomingCountdown(params?: PaginationParams): Promise<any> {
    const response = await api.get('/games/upcoming-countdown', { params });
    return response.data;
  },

  async getReleaseTimeline(params?: PaginationParams & { type?: string }): Promise<any> {
    const response = await api.get('/games/release-timeline', { params });
    return response.data;
  },

  async getGameCountdown(gameId: number): Promise<any> {
    const response = await api.get(`/games/${gameId}/countdown`);
    return response.data;
  },

  async getAbandonwareGames(params?: PaginationParams): Promise<PaginatedResponse<Game>> {
    const response = await api.get<PaginatedResponse<Game>>('/games/abandonware', { params });
    return response.data;
  },
};

File: mobile-app/lib/services/game_service.dart
import '../models/game.dart';
import 'api_service.dart';

class PaginatedGames {
  final List<Game> games;
  final int total;
  final int page;
  final int totalPages;

  PaginatedGames({
    required this.games,
    required this.total,
    required this.page,
    required this.totalPages,
  });

  bool get hasMore => page < totalPages;
}

class GameService {
  final ApiService _api = ApiService();

  /// Fetch all games with optional pagination, search and filters.
  Future<PaginatedGames> getAllGames({
    int page = 1,
    int limit = 20,
    String? search,
    String? releaseStatus,
    String? availabilityStatus,
    int? year,
    String? tag,
    String? platform,
    String? genre,
    String? sort,
  }) async {
    final params = <String, String>{
      'page': page.toString(),
      'limit': limit.toString(),
    };
    if (search != null && search.isNotEmpty) params['search'] = search;
    if (releaseStatus != null) params['release_status'] = releaseStatus;
    if (availabilityStatus != null) params['availability_status'] = availabilityStatus;
    if (year != null) params['year'] = year.toString();
    if (tag != null) params['tag'] = tag;
    if (platform != null) params['platform'] = platform;
    if (genre != null) params['genre'] = genre;
    if (sort != null) params['sort'] = sort;

    final response = await _api.get('/games', queryParams: params);
    final data = response['data'] as Map<String, dynamic>;
    final pagination = data['pagination'] as Map<String, dynamic>? ??
        response['pagination'] as Map<String, dynamic>? ??
        {};

    final gamesJson = data['games'] as List<dynamic>? ?? [];
    final games = gamesJson
        .map((g) => Game.fromJson(g as Map<String, dynamic>))
        .toList();

    return PaginatedGames(
      games: games,
      total: pagination['total'] as int? ?? games.length,
      page: pagination['page'] as int? ?? page,
      totalPages: pagination['totalPages'] as int? ?? 1,
    );
  }

  /// Get a single game by ID.
  Future<Game> getGameById(int id) async {
    final response = await _api.get('/games/$id');
    final data = response['data'] as Map<String, dynamic>;
    return Game.fromJson(data['game'] as Map<String, dynamic>? ?? data);
  }

  /// Search games by text query.
  Future<PaginatedGames> searchGames({
    required String query,
    int page = 1,
    int limit = 20,
  }) async {
    final response = await _api.get('/games/search', queryParams: {
      'q': query,
      'page': page.toString(),
      'limit': limit.toString(),
    });

    final data = response['data'] as Map<String, dynamic>;
    final pagination = data['pagination'] as Map<String, dynamic>? ?? {};
    final gamesJson = data['games'] as List<dynamic>? ?? [];
    final games = gamesJson
        .map((g) => Game.fromJson(g as Map<String, dynamic>))
        .toList();

    return PaginatedGames(
      games: games,
      total: pagination['total'] as int? ?? games.length,
      page: pagination['page'] as int? ?? page,
      totalPages: pagination['totalPages'] as int? ?? 1,
    );
  }

  /// Get upcoming game releases.
  Future<PaginatedGames> getUpcomingReleases({int page = 1, int limit = 20}) async {
    final response = await _api.get('/games/upcoming-releases', queryParams: {
      'page': page.toString(),
      'limit': limit.toString(),
    });

    final data = response['data'] as Map<String, dynamic>;
    final pagination = data['pagination'] as Map<String, dynamic>? ?? {};
    final gamesJson = data['games'] as List<dynamic>? ?? [];
    final games = gamesJson
        .map((g) => Game.fromJson(g as Map<String, dynamic>))
        .toList();

    return PaginatedGames(
      games: games,
      total: pagination['total'] as int? ?? games.length,
      page: pagination['page'] as int? ?? page,
      totalPages: pagination['totalPages'] as int? ?? 1,
    );
  }

  /// Get abandonware games.
  Future<PaginatedGames> getAbandonwareGames({int page = 1, int limit = 20}) async {
    final response = await _api.get('/games/abandonware', queryParams: {
      'page': page.toString(),
      'limit': limit.toString(),
    });

    final data = response['data'] as Map<String, dynamic>;
    final pagination = data['pagination'] as Map<String, dynamic>? ?? {};
    final gamesJson = data['games'] as List<dynamic>? ?? [];
    final games = gamesJson
        .map((g) => Game.fromJson(g as Map<String, dynamic>))
        .toList();

    return PaginatedGames(
      games: games,
      total: pagination['total'] as int? ?? games.length,
      page: pagination['page'] as int? ?? page,
      totalPages: pagination['totalPages'] as int? ?? 1,
    );
  }

  /// Get GOTY games.
  Future<PaginatedGames> getGotyGames({int page = 1, int limit = 20}) async {
    final response = await _api.get('/games/goty', queryParams: {
      'page': page.toString(),
      'limit': limit.toString(),
    });

    final data = response['data'] as Map<String, dynamic>;
    final pagination = data['pagination'] as Map<String, dynamic>? ?? {};
    final gamesJson = data['games'] as List<dynamic>? ?? [];
    final games = gamesJson
        .map((g) => Game.fromJson(g as Map<String, dynamic>))
        .toList();

    return PaginatedGames(
      games: games,
      total: pagination['total'] as int? ?? games.length,
      page: pagination['page'] as int? ?? page,
      totalPages: pagination['totalPages'] as int? ?? 1,
    );
  }
}
